bucket replicate panics on prometheus registration

<!--
Template relevant to bug reports only!

Keep issue title verbose enough and add prefix telling
about what components it touches e.g "query:" or ".*:"
-->

<!--
In case of issues related to exact bucket implementation, please ping corresponded maintainer from list here: https://github.com/thanos-io/thanos/blob/master/docs/storage.md
-->

**Thanos, Prometheus and Golang version used**: 
thanosio/thanos:master-2020-03-10-50ffaaaa

<!--
Output of "thanos --version" or docker image:tag used.
(Double-check if all deployed components/services have expected versions)

If you are using custom build from master branch, have you checked out the tip of the master?
-->

**Object Storage Provider**: 
S3

**What happened**:
When attempting to replicate data between two S3 buckets, a panic is reported stating `duplicate metrics collector registration attempted`

**What you expected to happen**:
The process should replicate data without panicking

**How to reproduce it (as minimally and precisely as possible)**:
Ran the following command in a docker environment:
```
bucket replicate --objstore.config-file=/config/from-config.yaml --objstore-to.config-file=/config/to-config.yaml --log.level=debug --single-run
```
Where the config files point to two S3 buckets in the same AWS region
**Full logs to relevant components**:

<details>Logs
<p>

```
level=debug ts=2020-03-10T14:53:52.496252139Z caller=main.go:104 msg="maxprocs: Leaving GOMAXPROCS=[6]: CPU quota undefined"
level=info ts=2020-03-10T14:53:52.496444563Z caller=main.go:153 msg="Tracing will be disabled"
level=debug ts=2020-03-10T14:53:52.496470735Z caller=replicater.go:78 component=replicate msg="setting up http listen-group"
level=info ts=2020-03-10T14:53:52.498942801Z caller=factory.go:46 component=replicate msg="loading bucket configuration"
level=info ts=2020-03-10T14:53:52.501258682Z caller=factory.go:46 component=replicate msg="loading bucket configuration"
level=info ts=2020-03-10T14:53:52.501665456Z caller=replicater.go:208 component=replicate msg="starting replication"
level=info ts=2020-03-10T14:53:52.501793108Z caller=replicater.go:92 component=replicate msg="Listening for http service" address=0.0.0.0:10902
level=info ts=2020-03-10T14:53:52.501966695Z caller=intrumentation.go:60 component=replicate msg="changing probe status" status=healthy
level=info ts=2020-03-10T14:53:52.502053157Z caller=http.go:56 component=replicate service=http/server component=replicate msg="listening for requests and metrics" address=0.0.0.0:10902
level=info ts=2020-03-10T14:53:52.502134293Z caller=replicater.go:171 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D msg="running replication attempt"
level=debug ts=2020-03-10T14:53:52.502186639Z caller=scheme.go:173 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D msg="scanning blocks available blocks for replication"
level=debug ts=2020-03-10T14:53:53.427112952Z caller=fetcher.go:358 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D component=block.MetaFetcher msg="successfully fetched block metadata" duration=525.196648ms cached=14 returned=14 partial=0
level=debug ts=2020-03-10T14:53:53.427249937Z caller=scheme.go:204 component=replicate replication-run-id=01E32F04FNKG36J2NXH571FJ3D msg="adding block to available blocks" block_uuid=01E1TMA6ABM79ENM26R8NQSJ6A
panic: duplicate metrics collector registration attempted

goroutine 131 [running]:
github.com/prometheus/client_golang/prometheus.(*Registry).MustRegister(0xc0001fea00, 0xc000a38430, 0x1, 0x1)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.5.0/prometheus/registry.go:400 +0xad
github.com/prometheus/client_golang/prometheus/promauto.Factory.NewCounter(0x22c42e0, 0xc0001fea00, 0x0, 0x0, 0x1ea614a, 0xb, 0x1ea6bf2, 0xb, 0x1ef4f43, 0x2e, ...)
	/go/pkg/mod/github.com/prometheus/client_golang@v1.5.0/prometheus/promauto/auto.go:263 +0x12f
github.com/thanos-io/thanos/pkg/block.newSyncMetrics(0x22c42e0, 0xc0001fea00, 0xc00096f4a0)
	/go/src/github.com/thanos-io/thanos/pkg/block/fetcher.go:61 +0xa9
github.com/thanos-io/thanos/pkg/block.NewMetaFetcher(0x228a7e0, 0xc000ab2030, 0x20, 0x7f33dad9d0a0, 0xc000a791d0, 0x0, 0x0, 0x22c42e0, 0xc0001fea00, 0x0, ...)
	/go/src/github.com/thanos-io/thanos/pkg/block/fetcher.go:139 +0x10e
github.com/thanos-io/thanos/pkg/replicate.loadMeta(0x22d95a0, 0xc000156680, 0xc0000d80a0, 0x37d590d22da97001, 0x177a7b40c0b9de22, 0xc000138390, 0x0, 0x1, 0x0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/scheme.go:353 +0x8f
github.com/thanos-io/thanos/pkg/replicate.(*replicationScheme).execute.func1(0xc000c32140, 0x1b, 0x1, 0x0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/scheme.go:185 +0xfd
github.com/thanos-io/thanos/pkg/objstore/s3.(*Bucket).Iter(0xc0001fecd0, 0x22d95a0, 0xc000156680, 0x0, 0x0, 0xc000ab2180, 0x30, 0x1ce0760)
	/go/src/github.com/thanos-io/thanos/pkg/objstore/s3/s3.go:243 +0x2c0
github.com/thanos-io/thanos/pkg/objstore.(*metricBucket).Iter(0xc000a791d0, 0x22d95a0, 0xc000156680, 0x0, 0x0, 0xc000ab2180, 0xc000ab2120, 0x8)
	/go/src/github.com/thanos-io/thanos/pkg/objstore/objstore.go:250 +0x6f
github.com/thanos-io/thanos/pkg/replicate.(*replicationScheme).execute(0xc0000d80a0, 0x22d95a0, 0xc000156680, 0x7f33dad9d0a0, 0xc000a791d0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/scheme.go:175 +0x1f4
github.com/thanos-io/thanos/pkg/replicate.RunReplicate.func3(0xc00038de38, 0x2307da0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/replicater.go:173 +0x609
github.com/thanos-io/thanos/pkg/replicate.RunReplicate.func4(0x0, 0x0)
	/go/src/github.com/thanos-io/thanos/pkg/replicate/replicater.go:185 +0x23c
github.com/oklog/run.(*Group).Run.func1(0xc00097c1e0, 0xc00003b730, 0xc000a7d460)
	/go/pkg/mod/github.com/oklog/run@v1.0.0/group.go:38 +0x27
created by github.com/oklog/run.(*Group).Run
	/go/pkg/mod/github.com/oklog/run@v1.0.0/group.go:37 +0xbb
```

</p>
</details>

<!--
Uncomment and fill if you use not casual environment or if it might be relevant.

**Environment**:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Others:

-->

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@ We use *breaking* word for marking changes that are not backward compatible (rel
 
 - [#2238](https://github.com/thanos-io/thanos/pull/2238) Ruler: Fixed Issue #2204 bug in alert queue signalling filled up queue and alerts were dropped
 - [#2231](https://github.com/thanos-io/thanos/pull/2231) Bucket Web - Sort chunks by thanos.downsample.resolution for better grouping
+- [#2254](https://github.com/thanos-io/thanos/pull/2254) Bucket: Fix metrics registered multiple times in bucket replicate
 
 ### Added
 

--- a/pkg/replicate/replicator.go
+++ b/pkg/replicate/replicator.go
@@ -20,6 +20,7 @@ import (
 	"github.com/prometheus/client_golang/prometheus/promauto"
 	"github.com/prometheus/common/model"
 	"github.com/prometheus/prometheus/pkg/labels"
+	thanosblock "github.com/thanos-io/thanos/pkg/block"
 	"github.com/thanos-io/thanos/pkg/compact"
 	"github.com/thanos-io/thanos/pkg/component"
 	"github.com/thanos-io/thanos/pkg/extflag"
@@ -149,6 +150,11 @@ func RunReplicate(
 		Help: "The Duration of replication runs split by success and error.",
 	}, []string{"result"})
 
+	fetcher, err := thanosblock.NewMetaFetcher(logger, 32, fromBkt, "", reg)
+	if err != nil {
+		return errors.Wrapf(err, "create meta fetcher with bucket %v", fromBkt)
+	}
+
 	blockFilter := NewBlockFilter(
 		logger,
 		labelSelector,
@@ -170,7 +176,7 @@ func RunReplicate(
 		logger := log.With(logger, "replication-run-id", ulid.String())
 		level.Info(logger).Log("msg", "running replication attempt")
 
-		if err := newReplicationScheme(logger, metrics, blockFilter, fromBkt, toBkt, reg).execute(ctx); err != nil {
+		if err := newReplicationScheme(logger, metrics, blockFilter, fetcher, fromBkt, toBkt, reg).execute(ctx); err != nil {
 			return errors.Wrap(err, "replication execute")
 		}
 

--- a/pkg/replicate/scheme.go
+++ b/pkg/replicate/scheme.go
@@ -105,6 +105,7 @@ type replicationScheme struct {
 	toBkt   objstore.Bucket
 
 	blockFilter blockFilterFunc
+	fetcher     thanosblock.MetadataFetcher
 
 	logger  log.Logger
 	metrics *replicationMetrics
@@ -152,7 +153,15 @@ func newReplicationMetrics(reg prometheus.Registerer) *replicationMetrics {
 	return m
 }
 
-func newReplicationScheme(logger log.Logger, metrics *replicationMetrics, blockFilter blockFilterFunc, from objstore.BucketReader, to objstore.Bucket, reg prometheus.Registerer) *replicationScheme {
+func newReplicationScheme(
+	logger log.Logger,
+	metrics *replicationMetrics,
+	blockFilter blockFilterFunc,
+	fetcher thanosblock.MetadataFetcher,
+	from objstore.BucketReader,
+	to objstore.Bucket,
+	reg prometheus.Registerer,
+) *replicationScheme {
 	if logger == nil {
 		logger = log.NewNopLogger()
 	}
@@ -160,6 +169,7 @@ func newReplicationScheme(logger log.Logger, metrics *replicationMetrics, blockF
 	return &replicationScheme{
 		logger:      logger,
 		blockFilter: blockFilter,
+		fetcher:     fetcher,
 		fromBkt:     from,
 		toBkt:       to,
 		metrics:     metrics,
@@ -350,12 +360,7 @@ func (rs *replicationScheme) ensureObjectReplicated(ctx context.Context, objectN
 // partial, this is just a temporary failure, as the block is still being
 // uploaded to the origin bucket.
 func loadMeta(ctx context.Context, rs *replicationScheme, id ulid.ULID) (*metadata.Meta, bool, error) {
-	fetcher, err := thanosblock.NewMetaFetcher(rs.logger, 32, rs.fromBkt, "", rs.reg)
-	if err != nil {
-		return nil, false, errors.Wrapf(err, "create meta fetcher with buecket %v", rs.fromBkt)
-	}
-
-	metas, _, err := fetcher.Fetch(ctx)
+	metas, _, err := rs.fetcher.Fetch(ctx)
 	if err != nil {
 		switch errors.Cause(err) {
 		default:

--- a/pkg/replicate/scheme_test.go
+++ b/pkg/replicate/scheme_test.go
@@ -19,6 +19,7 @@ import (
 	"github.com/oklog/ulid"
 	"github.com/prometheus/prometheus/pkg/labels"
 	"github.com/prometheus/prometheus/tsdb"
+	"github.com/thanos-io/thanos/pkg/block"
 	"github.com/thanos-io/thanos/pkg/block/metadata"
 	"github.com/thanos-io/thanos/pkg/compact"
 	"github.com/thanos-io/thanos/pkg/objstore"
@@ -302,9 +303,7 @@ func TestReplicationSchemeAll(t *testing.T) {
 		c.prepare(ctx, t, originBucket, targetBucket)
 
 		matcher, err := labels.NewMatcher(labels.MatchEqual, "test-labelname", "test-labelvalue")
-		if err != nil {
-			t.Fatal("Failed to create a matcher.")
-		}
+		testutil.Ok(t, err)
 
 		selector := labels.Selector{
 			matcher,
@@ -314,11 +313,14 @@ func TestReplicationSchemeAll(t *testing.T) {
 		}
 
 		filter := NewBlockFilter(logger, selector, compact.ResolutionLevelRaw, 1).Filter
+		fetcher, err := block.NewMetaFetcher(logger, 32, originBucket, "", nil)
+		testutil.Ok(t, err)
 
 		r := newReplicationScheme(
 			logger,
 			newReplicationMetrics(nil),
 			filter,
+			fetcher,
 			originBucket,
 			targetBucket,
 			nil,

